pkg/xds/generator/direct_access_proxy_generator.go

```diff
@@ -0,0 +1,146 @@
+package generator
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/pkg/errors"
+
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+	core_xds "github.com/Kong/kuma/pkg/core/xds"
+	xds_context "github.com/Kong/kuma/pkg/xds/context"
+	envoy_common "github.com/Kong/kuma/pkg/xds/envoy"
+	envoy_clusters "github.com/Kong/kuma/pkg/xds/envoy/clusters"
+	envoy_listeners "github.com/Kong/kuma/pkg/xds/envoy/listeners"
+)
+
+// Transparent Proxy is based on having 1 IP for cluster (ex. ClusterIP of Service on K8S), so consuming apps by their IP
+// is unknown destination from Envoy perspective. Therefore such request will go trough pass_trough cluster and won't be encrypted by mTLS.
+// This generates listener for every IP and redirect traffic trough "direct_access" cluster which is configured to encrypt connections.
+// Generating listener for every endpoint will cause XDS snapshot to be large therefore it should be used only if really needed.
+//
+// Second approach to consider was to use FilterChainMatch on catch_all listener with list of all direct access endpoints
+// instead of generating outbound listener, but it seemed to not work with Listener#UseOriginalDst
```

Comment:
There is a comment in envoy docs about `use_original_dst`:

_This field is deprecated. Use an original_dst listener filter instead.
Note that hand off to another listener is NOT performed without this flag. Once FilterChainMatch is implemented this flag will be removed, as filter chain matching can be used to select a filter chain based on the restored destination address._

So probably the second approach is workable as well.

Reply:
Yes, I know. I've seen this. I tried to get rid of `use_original_dst` in this PR, but it requires changes in iptables, permissions on K8S, and other stuff. I didn't manage get this to work.